### ISSUE TITLE
TestNG marks the cucumber failed scenarios as SKIP

### ISSUE THREAD

**Issue by gopi-accnt** [cbeust/testng#2742]
### TestNG Version
7.3.0

### Expected behavior
TestNG to mark the Cucumber failed scenario as FAILURE and trigger retry.

### Actual behavior
TestNG marks the Cucumber failed scenario as SKIP and so retry is not getting triggered.

### Is the issue reproducible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

### Test case sample
I have been using TestNG 6.8.0 version till before and now trying to upgrade it to TestNG 7.3.0. But while trying to do so, my retry logic is not working. I have written Cucumber + TestNG based test runner class where I have custom test listener that enables retry upon test failures. But after this upgrade, TestNG is marking the failed Cucumber scenarios as SKIPPED which eventually impacts the retry mechanism.

```
@CucumberOptions(
        features = "src/main/java/Features",
        glue = {"stepDefinitions"},
        tags = {"~@Ignore"},
        format = {"pretty","json:target/cucumber-reports/CucumberTestReport.json"},
        plugin = "json:target/cucumber-reports/CucumberTestReport.json")

@Listener("RetryListener.class")
public class TestRunner {
   
    private TestNGCucumberRunner testNGCucumberRunner;
    
    @BeforeClass(alwaysRun = true)
    public void setUpCucumber() {
         testNGCucumberRunner = new TestNGCucumberRunner(this.getClass());
    }

    @Test(groups = "cucumber", description = "Runs Cucumber Feature", dataProvider = "features", retryAnalyzer="RetryAnalyzer.class")
    public void feature(CucumberFeatureWrapper cucumberFeature) throws Throwable {
        testNGCucumberRunner.runCucumber(cucumberFeature.getCucumberFeature());
    }
 
    @DataProvider
    public Object[][] features() {
        return testNGCucumberRunner.provideFeatures();
    }
 
    @AfterClass(alwaysRun = true)
    public void tearDownClass() throws Exception {
        testNGCucumberRunner.finish();
    }
}
```

Note: Due to some restrictions I am forced to use 7.3.0 only, so it will be great if someone can share any idea on how to resolve this issue with this version itself.

**Comment by krmahadevan** [comment]
@gopi-accnt please try this using 7.5 version. 

Even if the problem exists in 7.3.0 there is not much that can be done. 

Please post back the results of trying this out on 7.5

**Comment by krmahadevan** [comment]
@gopi-accnt if the problem persists on 7.5 then please share a simple example ( without cucumber integration) that can be used to reproduce the problem.

**Comment by gopi-accnt** [comment]
@krmahadevan thanks for the quick response!

I am facing same issue with the latest version of TestNG too. I am afraid if I can exactly simplify the problem without cucumber integration, but I have noticed the difference.. i.e., the way test status being marked between 6.8.x and the latest version. Hopefully this could give you some background.

Once the test failed, Cucumber retrieves the exception as `Throwable` and sends to TestNG invoker for updating the test status from SKIP to FAILURE. Until 6.8.x it was simply overriding the test status as FAILURE, but now it is doing it only if the test status is STARTED.

This is the [line](https://github.com/cbeust/testng/blob/master/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java#L516-L522) where Cucumber throw reaches after retrieving the test exception upon failure.

### Latest
```
private static void setTestStatus(ITestResult result, int status) {
    // set the test to success as long as the testResult hasn't been changed by the user via
    // Reporter.getCurrentTestResult
    if (result.getStatus() == ITestResult.STARTED) {
      result.setStatus(status);
    }
  }
```
Earlier it was simply like below in `Invoker.class`

### 6.8.x

```
catch (InvocationTargetException var25) {
            testResult.setThrowable(var25.getCause());
            testResult.setStatus(2);
        }
```
Please let me know if still any additional information is needed.

**Comment by krmahadevan** [comment]
@gopi-accnt if the issue persists in 7.5 then please help share a simple sample ( without any third party libraries) that can be used to reproduce the issue. 

Without a standalone sample its not going to be possible to find out what went wrong and we would really like to not get distracted by what other libraries are doing in conjunction with TestNG which usually makes the debugging process a bit more involved and time consuming. 

Thanks in advance for your understanding.

**Comment by krmahadevan** [comment]
Closing this issue since there's no response for the request for a sample. Please add a comment to this issue once there's a sample that can be used to reproduce the problem using latest released version of TestNG (`7.5` as of today)
